### ISSUE TITLE
no such file or directory in: ${file("~/.ssh/id_rsa.pub")}

### ISSUE THREAD

**Issue by yuriko-h** [appsecco/breaking-and-pwning-apps-and-servers-aws-azure-training#2]
Hi,

Hoping you can assist with the below issue I've run into.

I was able to to run the following commands successfully:
eval `ssh-agent -s`
ssh-checker
ssh-add -l

But then received the below error when running 'deploy-cloudhacker':
Error: aws_key_pair.cloudhackerlabkey: 1 error(s) occurred:

* aws_key_pair.cloudhackerlabkey: file: open /home/student/snap/terraform/216/.ssh/id_rsa.pub: no such file or directory in:

${file("~/.ssh/id_rsa.pub")}

Thank you.

**Comment by riyazwalikar** [comment]
For some weird reason, the terraform installed using snap seems to be using a different home directory than the actual /home/student/.ssh/ folder.

I have pushed a change to the cloudhacker.tf file that should fix this. Pull and check again?

**Comment by yuriko-h** [comment]
> For some weird reason, the terraform installed using snap seems to be using a different home directory than the actual /home/student/.ssh/ folder.
> 
> I have pushed a change to the cloudhacker.tf file that should fix this. Pull and check again?

Hi thanks for the quick response.
Slightly different error this time.

* aws_key_pair.cloudhackerlabkey: file: open /home/student/.ssh/id_rsa.pub: permission denied in:

${file("/home/student/.ssh/id_rsa.pub")}

**Comment by yuriko-h** [comment]
> > For some weird reason, the terraform installed using snap seems to be using a different home directory than the actual /home/student/.ssh/ folder.
> > I have pushed a change to the cloudhacker.tf file that should fix this. Pull and check again?
> 
> Hi thanks for the quick response.
> Slightly different error this time.
> 
>     * aws_key_pair.cloudhackerlabkey: file: open /home/student/.ssh/id_rsa.pub: permission denied in:
> 
> 
> ${file("/home/student/.ssh/id_rsa.pub")}

Sorry accidentally closed it

**Comment by riyazwalikar** [comment]
Hi,

Can you let me know the following?

1. What user are you running the deploy-cloudhacker command
2. What are the permissions on the /home/student/.ssh folder
3. What are the permissions on the /home/student/.ssh/id_rs.pub

**Comment by angrypapa** [comment]
The problem seems to be the syntax, not the perms. I just changed it to:

```
public_key = "ssh-rsa AAA....."
```

And it works... well, not really because then I get: 

```
Error: aws_instance.attacker-machine: 1 error(s) occurred:

* aws_instance.attacker-machine: file: open /home/student/snap/terraform/216/.ssh/id_rsa: no such file or directory in:

${file("~/.ssh/id_rsa")}
```

Which seems related to your previous fix... but I can't find this anywhere in the config files... 

I noticed it creates the following empty dir in the student home directory `snap/terraform/216/`. Where is this coming from?

Thank you.

**Comment by angrypapa** [comment]
Copying `.ssh` to /home/student/snap/terraform/216` solves the problem, but then I get: 

```
provider.aws: No valid credential sources found for AWS Provider.
```

**Comment by angrypapa** [comment]
Exporting the keys as documented here worked for me https://www.terraform.io/docs/providers/aws/index.html

**Comment by angrypapa** [comment]
Guess I spoke too early, at the end I get:

```
Error: Error applying plan:

1 error(s) occurred:

* aws_instance.attacker-machine: Error connecting to SSH_AUTH_SOCK: dial unix /tmp/ssh-sJKsrBqf63X3/agent.24249: connect: no such file or directory
````

**Comment by angrypapa** [comment]
Solved it by adding `agent = "false"` to the connection conf directive but then...

```
Error: Error applying plan:

1 error(s) occurred:

* aws_instance.attacker-machine: Upload failed: open /home/student/.aws: permission denied
```

:]

**Comment by angrypapa** [comment]
Really appreciate help with this "last" one, as my google fu is not helping. 

From what I can see installing terraform via snap changes the "root" folder to `~/snap/terraform/216`. The way snap installs the binary seems to break integration with ssh-agent. I tried to create/copy the `.aws` directory to `~/snap/terraform/216` but no luck.

**Comment by riyazwalikar** [comment]
Terraform at some point in time deprecated quoted type constraints and couple of other things. I tested this again with terraform version 0.11.3 and was able to complete the installation.

I have made a new commit with the updated instructions to install terraform. If you have installed terraform using snap, then remove it using `sudo snap remove terraform` and use the updated instructions in `documentation/setting-up/setup-student-virtual-machine.md` to install the 0.11.3 version for the plans in this repo. I will try and update the plan files at a later point in time.
